Add fasttext as an option alongside cld2

jelmervdl opened the pull request:
This takes @kpu's work on using fasttext but adds it as an option alongside cld2. To make it easier to compare the two (or more, if we integrate more classifiers).

Currently in draft just to notify other followers of this work. The cld2 code path triggers a EXC_BAD_ACCESS at some point when writing to the url.gz file? It's a bit of a mystery, hence draft.

Edit for posterity:
```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x68)
    frame #0: 0x00007ff80c8d6d13 libsystem_c.dylib`flockfile + 18
libsystem_c.dylib`flockfile:
->  0x7ff80c8d6d13 <+18>: movq   0x68(%rbx), %rdi
    0x7ff80c8d6d17 <+22>: addq   $0x8, %rdi
    0x7ff80c8d6d1b <+26>: callq  0x7ff80c94691e            ; symbol stub for: pthread_mutex_lock
    0x7ff80c8d6d20 <+31>: callq  0x7ff80c94643e            ; symbol stub for: __error
Target 0: (warc2text) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x68)
  * frame #0: 0x00007ff80c8d6d13 libsystem_c.dylib`flockfile + 18
    frame #1: 0x00007ff80c8ebaea libsystem_c.dylib`fwrite + 66
    frame #2: 0x00000001000f0991 warc2text`warc2text::GzipWriter::compress(this=0x0000600003515db8, in="https://mt.wikipedia.org/wiki/Spe%C4%8Bjali:Statistika", size=54, flush=0) at bilangwriter.cc:43:13
    frame #3: 0x00000001000f0b20 warc2text`warc2text::GzipWriter::writeLine(this=0x0000600003515db8, text="https://mt.wikipedia.org/wiki/Spe%C4%8Bjali:Statistika") at bilangwriter.cc:65:15
    frame #4: 0x00000001000f0ff0 warc2text`warc2text::BilangWriter::write(this=0x00007ff7bfefe690, lang="mlt_Latn", b64text="U3RhdGlzdGlrYSAtIFdpa2lwZWRpamEKU3RhdGlzdGlrYQpBcWJlxbwgbGVqbjogbmF2aWdhenpqb25pLCBmaXR0ZXgKU3RhdGlzdGlrYSByZWxhdGF0YSBtYWwtcGHEoW5pClBhxKFuaSB0YScga29udGVudXQKMywyNDcKUGHEoW5pCihJbC1wYcShbmkga29sbGhhIGZpbC13aWtpLCBsaSBqaW5rbHVkdSBwYcShbmkgdGEnIGRpc2t1c3Nqb25pLCByaWluZGlyaXp6YW1lbnRpLCBldMSLLikKMTUsMDI0CkZhamxzIG10ZWxsYScKMSwxODUKU3RhdGlzdGlrYSB0YWwtaW1tb2RpZmlrYXIKTW9kaWZpa2kgbWlubiBtaW5kdSBsLVdpa2lwZWRpamEgYmRpZXQgaWwtZnVuempvbiB0YWfEp2hhCjI2MywwMTAKTWVkamEgdGEnIG1vZGlmaWtpIGt1bGwgcGHEoW5hCjE3LjUxClN0YXRpc3Rpa2EgdGFsLXV0ZW50ClV0ZW50aSByZcShaXN0cmF0aQoxMyw3NjYKVXRlbnRpIGF0dGl2aSAobGlzdGEgdGEnIG1lbWJyaSkKKFV0ZW50aSBsaSB3ZXR0cXUgYXp6am9uaSBmbC1hxKfEp2FyIDMwIGp1bSkKMjIKQm90IChsaXN0YSB0YScgbWVtYnJpKQo1MQpBbW1pbmlzdHJhdHVyaSAobGlzdGEgdGEnIG1lbWJyaSkKMwpCdXJva3JhdGkgKGxpc3RhIHRhJyBtZW1icmkpCjIKU3R3ZXR0aSAobGlzdGEgdGEnIG1lbWJyaSkKMApLcmVhdHVyaSB0YWwta29udGlqaWV0IChsaXN0YSB0YScgbWVtYnJpKQowCkltcG9ydGF0dXJpIChsaXN0YSB0YScgbWVtYnJpKQowCkltcG9ydGF0dXJpIHRyYW5zd2lraSAobGlzdGEgdGEnIG1lbWJyaSkKMAplxbxlbnpqb25pamlldCB0YWwt"..., url="https://mt.wikipedia.org/wiki/Spe%C4%8Bjali:Statistika", mime="text/html", b64html="") at bilangwriter.cc:90:16
    frame #5: 0x00000001000f1a35 warc2text`warc2text::BilangWriter::write(this=0x00007ff7bfefe690, record=0x00007ff7bfefe0f8, multilang=true, paragraph_identification=false) at bilangwriter.cc:128:23
    frame #6: 0x00000001000388e5 warc2text`warc2text::WARCPreprocessor::process(this=0x00007ff7bfefe688, filename="WIDE-20171230055008-00094.warc.gz") at warcpreprocessor.cc:164:20
    frame #7: 0x000000010000a861 warc2text`main(argc=8, argv=0x00007ff7bfefeeb0) at warc2text_main.cc:129:19
    frame #8: 0x00007ff80c6c9310 dyld`start + 2432
```

jelmervdl commented:
Not a programming error, just a *too many files* open.

lpla commented:
I think that someone used this "unk" mark to process WARC files in Bitextor that contain a CLD2 unsupported language (only one, of course). So I would say to keep this behaviour as it is now.

lpla commented:
I think so. At least, it seems like at some point that worked somehow: https://github.com/bitextor/bitextor/issues/109#issuecomment-555842143

jelmervdl commented:
I've changed it now so that any record that cannot reliably be classified as a certain language is labelled as "unk". This matches the behaviour that warc2text with CLD2 already used to have.

So `unk/` will contain all records that:
- can either not be reliably be classified as one language (or multiple, if `--multilang`),
- or are reliably definitely `unk` according to CLD2.
- FastText models don't seem to have a specific label for `unk`, so there it is just anything that doesn't match the threshold (0.5 hardcoded, not sure whether that should be a command line argument).

If you concat all the `output_folder/*/text.gz` records together, you should once again get all records in the warc, regardless of what the language identifier thought of them. Records that got filtered by the content-type, url filter or url extension filter will still be omitted.

I noticed the `reliable` boolean that got returned wasn't being used, so I removed it. I've also updated the counts in the log files to not count `unk` as "language bytes".

lpla commented:
Great! Let us test the installation of your fork with fasttext in Bitextor and check that our tests still pass in there, and we will merge this.

aarongaliano commented:
Our tests still pass with the changes, but we have problems with the installation.
After the `make install` we get this error:
```
CMake Error at fasttext/cmake_install.cmake:46 (file):
  file INSTALL cannot find "/.../warc2text/build/fasttext.pc": No
  such file or directory.
Call Stack (most recent call first):
  cmake_install.cmake:93 (include)
```
This file is located at `/.../warc2text/build/fasttext/fasttext.pc`. If we edit the file `fasttext/cmake_install.cmake:46`  with the correct path, the installation works.

jelmervdl commented:
Aah, right, yes, sorry. I should have used `add_subdirectory(... EXLCUDE_FROM_ALL)`. I'll fix that.

ZJaume commented:
>     * FastText models don't seem to have a specific label for `unk`, so there it is just anything that doesn't match the threshold (0.5 hardcoded, not sure whether that should be a command line argument).
> 

We may give some thought to the use the threshold, because the default options on CLI and python are without a threshold. Fasttext just outputs the label that has higher probability. If we use threshold, we would be stricter than the cleaning tools and we may reduce coverage for low-resource langs. Uncertainty between the labels in fasttext, is something that happens frequently in low-resource and easily mistakeable languages (like some of the ones we address in [https://github.com/mbanon/fastspell/blob/main/src/fastspell/config/similar.yaml](fastspell)).

But yeah, we could always reprocess `unk`.
